### PR TITLE
Mnot/edits

### PR DESCRIPTION
**Pull request by mnot**
A proposed and rough rewrite. 

Feel free to merge, cherry-pick, or ignore -- but please keep the branch around for a little while so I can use it for reference.

### CODE REVIEW

File: draft-ietf-masque-h3-datagram.md [hunk 1]
```diff
@@ -28,65 +28,71 @@ author:
...
 When running over HTTP/2, demultiplexing is provided by the HTTP/2 framing
-layer. When running over HTTP/1, requests are strictly serialized in the
-connection, therefore demultiplexing is not needed.
+layer, but unreliable delivery is unavailable. HTTP Datagrams are negotiated
+and conveyed using the Capsule Protocol; see {{datagram-capsule}}.
+
+When running over HTTP/1, requests are strictly serialized in the connection,
+and therefore demultiplexing is not available. Unreliable delivery is likewise
+not available. HTTP Datagrams are negotiated and conveyed using the Capsule
+Protocol; see {{datagram-capsule}}.
```
Comment (philsbln) [inline draft-ietf-masque-h3-datagram.md:92]:
```suggestion
When running over HTTP/1, upgrading the connection to convay HTTP datagrams prevents other uses of the connection and therefore (de-)multiplexing not possible. Unreliable delivery is likewise
not available. HTTP Datagrams are negotiated and conveyed using the Capsule
Protocol; see {{datagram-capsule}}.
```
Closes #156

File: draft-ietf-masque-h3-datagram.md [hunk 2]
```diff
@@ -194,26 +207,114 @@ recent draft version from the intersection set. This ensures that both
...
-# Capsules {#capsule}
 
-This specification introduces the Capsule Protocol. The Capsule Protocol is a
-sequence of type-length-value tuples that new HTTP Upgrade Tokens (see {{Section
-16.7 of !HTTP=I-D.ietf-httpbis-semantics}}) can choose to use. It allows
-endpoints to reliably communicate request-related information end-to-end on HTTP
-request streams, even in the presence of HTTP intermediaries. The Capsule
-Protocol can be used to exchange HTTP Datagrams when HTTP is running over a
-transport that does not support the QUIC DATAGRAM frame.
+## The DATAGRAM Capsule {#datagram-capsule}
```
Comment (philsbln) [inline draft-ietf-masque-h3-datagram.md:211]: Closely match "HTTP/3 Datagrams" header
```suggestion
## HTTP/2 and HTTP/1.1 Datagrams using DATAGRAM Capsules {#datagram-capsule}
```

**Review by philsbln — APPROVED** [review]
This PR greatly improved readability and addresses my editorial nits of draft-ietf-masque-h3-datagram-07.

I made a few optional knit suggestions for more clarity.

File: draft-ietf-masque-h3-datagram.md [hunk 3]
```diff
@@ -28,65 +28,71 @@ author:
...
-application protocols running over QUIC {{!QUIC=RFC9000}} with a mechanism to
-send unreliable data while leveraging the security and congestion-control
-properties of QUIC. However, QUIC DATAGRAM frames do not provide a means to
-demultiplex application contexts. This document describes how to use QUIC
-DATAGRAM frames with HTTP/3 {{!H3=I-D.ietf-quic-http}} by association with HTTP
-requests. Additionally, this document defines the Capsule Protocol that can
-convey datagrams over prior versions of HTTP.
+HTTP extensions sometimes need to access underlying transport protocol features
+such as unreliable delivery (as offered by {{!DGRAM=I-D.ietf-quic-datagram}})
+to enable desirable features like an unreliable version of the CONNECT method,
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:56]:
```suggestion
to enable desirable features like an unreliable equivalent to the CONNECT method,
```

perhaps

File: draft-ietf-masque-h3-datagram.md [hunk 4]
```diff
@@ -28,65 +28,71 @@ author:
...
+and unreliable delivery in WebSockets {{?RFC6455}} (or its successors).
 
+In {{datagrams}}, this document describes HTTP Datagrams, a convention for
+conveying multiplexed, bidirectional datagrams inside a HTTP connection. They
```
[the remark this answers is not in the document]
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:60]: I'm not a huge fan of mixing terms like datagrams, streams, and streams of datagrams. That seems like a recipe for confusion and RFC editor purgatory.

File: draft-ietf-masque-h3-datagram.md [hunk 5]
```diff
@@ -28,65 +28,71 @@ author:
...
-to use QUIC DATAGRAM frames with HTTP/3 by association with HTTP requests.
-Additionally, this document defines the Capsule Protocol that can convey
-datagrams over prior versions of HTTP.
+This document describes HTTP Datagrams, a convention for conveying multiplexed,
+potentially unreliable datagrams inside an HTTP connection.
+
+In HTTP/3, HTTP Datagrams can be conveyed natively using the QUIC DATAGRAM
+extension. To enable seamless transfer of HTTP Datagrams in other protocol
+versions, this document also defines the Capsule Protocol, a more general
+convention for conveying data in HTTP connections.
```
[the remark this answers is not in the document]
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:45]: The version limitations are not strictly true. We would probably be better phrasing things, here and elsewhere in the document, more like "where QUIC DATAGRAMS are unavailable or undesireable (e.g. other HTTP versions or specific use cases) the Capsule Protocol provides a general convention for conveying data in HTTP connections, which can be used for datagram messages.

File: draft-ietf-masque-h3-datagram.md [hunk 6]
```diff
@@ -113,8 +119,8 @@ error of type H3_DATAGRAM_ERROR.
 
 HTTP Datagram Payload:
 
-: The payload of the datagram, whose semantics are defined by individual
-applications. Note that this field can be empty.
+: The payload of the datagram, whose semantics are defined by the
+application. Note that this field can be empty.
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:123]: In the abstract, you propose the addition of the text "Both are intended for use by HTTP extensions, not applications." do you think we need to, similarly, avoid using the term application here?

File: draft-ietf-masque-h3-datagram.md (hunk 2 above)
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:211]: again, this implies a version restriction the doesn't exist

File: draft-ietf-masque-h3-datagram.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (philsbln) [reply draft-ietf-masque-h3-datagram.md:60]: It is confusing… but I guess being vague about what HTTP datagrams are and how they relate to HTTP conversations does not ease that confusion.
Perhaps calling it "streams of datagrams" would make it a bit less confusing.

[the remark this answers is not in the document]
Comment (LPardue) [reply draft-ietf-masque-h3-datagram.md:60]: datagrams are datagrams. They are associated with request streams. How zero, one or many datagrams might get used in either direction is entirely up to the thing that uses them. 

Perhaps this paragraph would be better phrased by starting it along the lines of 

"In {{datagrams}}, this document describes HTTP datagrams, a convention that associates non-content data with HTTP requests to support the bidirectional, possibly multiplexed, exchange of datagrams inside an HTTP connection."

File: draft-ietf-masque-h3-datagram.md (hunk 1 above)
Comment (mnot) [reply draft-ietf-masque-h3-datagram.md:92]: This PR is intended to be purely editorial, so let's keep resolution of that separate.

File: draft-ietf-masque-h3-datagram.md (hunk 6 above)
Comment (mnot) [reply draft-ietf-masque-h3-datagram.md:123]: Possibly -- the semantics are definitely _contextualised_ by the extension in use, and it's probably up to it to determine whether they're delegated to the application.

how about

```suggestion
extension that is using HTTP Datagrams. Note that this field can be empty.
```

File: draft-ietf-masque-h3-datagram.md [hunk 7]
```diff
@@ -139,20 +145,27 @@ an error is not mandatory in this case because HTTP/3 implementations might have
...
-HTTP/3 datagrams MUST only be sent with an association to a stream that supports
-semantics for HTTP Datagrams. For example, existing HTTP methods GET and POST do
-not define semantics for associated HTTP Datagrams; therefore, HTTP/3 datagrams
-cannot be sent associated with GET or POST request streams. If an endpoint
-receives an HTTP/3 datagram associated with a method that has no known semantics
-for HTTP Datagrams, it MUST abort the corresponding stream with
-H3_DATAGRAM_ERROR. Future extensions MAY remove these requirements if they
-define semantics for such HTTP Datagrams and negotiate mutual support.
+HTTP/3 datagrams MUST only be sent with an association to a stream whose HTTP
+semantics explicitly supports HTTP Datagrams. For example, existing HTTP
```
Comment (mnot) [inline draft-ietf-masque-h3-datagram.md:149]: This is still a bit imprecise. suggest

```suggestion
HTTP/3 datagrams MUST only be sent with an association to a stream whose HTTP
semantics explicitly supports HTTP Datagrams (usually, in conjunction with the extension that invokes
the use of HTTP Datagrams. For example, existing HTTP
```

Also, these requirements seem general to all HTTP datagrams, but are placed against "HTTP/3 datagrams"

File: draft-ietf-masque-h3-datagram.md [hunk 8]
```diff
@@ -306,100 +383,29 @@ stream was truncated, this MUST be treated as a malformed or incomplete message.
...
-sending the Capsule-Protocol header field with a value of ?1. A Capsule-Protocol
-header field with a value of ?0 has the same semantics as when the header is not
-present. Intermediaries MAY use this header field to allow processing of HTTP
-Datagrams for unknown HTTP Upgrade Tokens; note that this is only possible for
-HTTP Upgrade or Extended CONNECT.
+Intermediaries MAY use this header field to allow processing of HTTP Datagrams
+for unknown HTTP Upgrade Tokens; note that this is only possible for HTTP
+Upgrade or Extended CONNECT.
 
-The Capsule-Protocol header field MUST NOT be sent multiple times on a message.
```
Comment (LPardue) [inline draft-ietf-masque-h3-datagram.md:324]: @mnot was there a reason to remove this multiplicity requirement? Can we track that as a separate issue?

**Comment by DavidSchinazi** [comment]
Closing as superseded by #158

**Pull request closed without merging**